src/repr/src/adt/apd.rs

```diff
@@ -83,15 +156,207 @@ pub fn twos_complement_be_to_i128(input: &[u8]) -> Result<i128, anyhow::Error> {
         // assigned the highest bit a weight of 128, instead of -128. To
         // compensate, we subtract twice the weight of the highest digit--
         // in the example, 256.
-        significand -= 2_i128.pow((input.len() * 8) as u32);
+        v -= 2_i128.pow((input.len() * 8) as u32);
     }
 
-    Ok(significand)
+    Ok(v)
+}
+
+pub fn apd_to_twos_complement_be(apd: Apd) -> [u8; ApdTwosComp::TWOS_COMPLEMENT_BYTE_WIDTH] {
+    let mut buf = [0; ApdTwosComp::TWOS_COMPLEMENT_BYTE_WIDTH];
+    let mut cx = ApdTwosComp::twos_comp_cx();
+    let mut apd = cx.to_width(apd);
+
+    // Ensure `apd` is a canonical coefficient.
+    if apd.exponent() < 0 {
+        let s = ApdTwosComp::from(-apd.exponent());
+        cx.scaleb(&mut apd, &s);
+    }
+
+    apd_to_twos_complement_inner::<ApdTwosComp, { APD_DATUM_WIDTH + 1 }>(apd, &mut cx, &mut buf);
+    buf
+}
+
+/// Provides a two's complement version of an APD appropriate for cases where a
+/// [`ScalarType::APD`]'s scale is unknown. This creates a value with 78 digits
+/// of precision and a scale of 39, which can definitively represent all valid
+/// [`Apd`] values.
+pub fn apd_to_twos_complement_wide(apd: Apd) -> [u8; ApdAgg::TWOS_COMPLEMENT_BYTE_WIDTH] {
+    let mut buf = [0; ApdAgg::TWOS_COMPLEMENT_BYTE_WIDTH];
+    let mut cx = ApdAgg::twos_comp_cx();
+    let mut d = cx.to_width(apd);
+    let mut scaler = ApdAgg::from(APD_DATUM_MAX_PRECISION);
+    cx.neg(&mut scaler);
+    // Shape `d` so that its exponent is -APD_DATUM_MAX_PRECISION
+    cx.rescale(&mut d, &scaler);
+    cx.abs(&mut scaler);
+    // Adjust `d` so it is a canonical coefficient, i.e. its exact value can be
+    // recovered by setting its exponent to -39.
+    cx.scaleb(&mut d, &scaler);
+
+    apd_to_twos_complement_inner::<ApdAgg, APD_AGG_WIDTH>(d, &mut cx, &mut buf);
+    buf
+}
+
+pub fn apd_to_twos_complement_inner<D: TwosCompDec<N>, const N: usize>(
```

Comment:
I'm going to try summarizing my understanding of what's going on here (because I'm almost certainly wrong about some of it):

1. Leading into the call to _inner you do some normalization/exponent futzing such that inner's sole job is to serialize the coefficient in 2's complement (plus some specified endianness, which I'll ignore since it's tangential)
2. Inside _inner you're doing the subtraction from 2^n, essentially using the math operators of the apd library to rearrange everything into 2's complement
3. Below in the while loop you're iterating over the coefficient's binary representation in chunks of u128 and writing those to a buffer, truncating early on the final chunk if we aren't actually using all the bits of the u128 for representation

Like you said, the 2^n deal requires a lot of hoops since you need one extra bit of space. Usually a good alternative is to instead invert the number in place and add 1 - but that relies on silent, ignored overflow which is... seems like exactly what apd is designed not to do. So unless apd exposes such a truncating overflow addition op, I'm guessing you'd be fighting the library. Does that sound accurate so far?

It seems like you miiiiight be able to do this conversion during iteration and drop the `inverter`, etc. 
I would break it up into two questions to see if it's tractable:
1. Is it possible to get a stream of bytes out of apd without too much math/creating extra Decimals such that if that stream were converted to 2's complement, it would be in the expected output format? I suspect this could be your while/rem/etc. loop, but also maybe iterating through `lsu` as-is would work (depending on that data format, I haven't looked at how things are stored in lsu).
2. Given 1, the problem's now 'can you convert a stream of bytes into 2's complement as you go?' Logically if you were going a bit at a time there's an alg (also mentioned on the wikipedia you linked) where you would emit 0s until you see the first 1, emit that 1, then flip every bit until the end. You can implement this sort of thing efficiently over steps larger than 1 bit (like a u64, u128, or whatever's convenient) by using things like the `clz` instruction (exposed in rust as `leading_zeroes`, https://en.wikipedia.org/wiki/Find_first_set) and masking to handle the initial portion, XOR 0xFF each chunk after to invert it, and apply a mask at the tail to truncate the final chunk to the total width limit.

For me the easiest way to iterate on that is by printing the raw binary representation of a concrete example that I'm trying to serialize next to the raw binary of the final encoded result and then working it out by hand a bit. I find it a lot harder to reason about if I don't visualize the representation at each step.

Reply:
> So unless apd exposes such a truncating overflow addition op, I'm guessing you'd be fighting the library.

Yeah, I don't believe that's an option because overflowing the representation doesn't cause it to cycle back around in the way primitive values do.

Re: using the `lsu`: the challenge with the approach using the underlying bytes is that they're all `u16` representing values in base 1,000--so you would need to somehow re-glue the representation of the `lsu`'s value into something that meaningfully saturated `u16`s (or something similar). Given the kind of inordinate complexity of moving from base 1,000 to base `65535`, I figured that doing this at the library level was simply more tractable, and that is more-or-less the current approach to getting byte-level representations of values.

I do think the approach using `leading_zeroes` is maybe more viable than I had originally thought. My hesitation was that it required an additional pass over the encoded array, but this really not performance-sensitive code--and that does pretty neatly tidy things up, e.g. letting me get rid of `inverter`.

tysm for taking a look at this!

Reply:
Ah, sadly there isn't a great way to avoid needing the extra digits of precision on the way from the byte representation into the `apd`. We're actually encoding a 17-byte number, which requires 41 digits of precision, which messes with this notion of having very large negative numbers and then adding unsigned bits to it to find our signed representation. So, as you suggested, I'm bailing on re-imagining this for the time being.